sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -130,14 +131,17 @@ case class InSubqueryExec(
     } else {
       rows.map(_.get(0, child.dataType))
     }
-    resultBroadcast = plan.session.sparkContext.broadcast(result)
+    if (needBroadcast) {
+      resultBroadcast = plan.session.sparkContext.broadcast(result)
+    }
   }
 
-  def values(): Option[Array[Any]] = Option(resultBroadcast).map(_.value)
+  // This is used only by DPP where we don't need broadcast the result.
+  def values(): Option[Array[Any]] = Option(result)
 
   private def prepareResult(): Unit = {
```

Comment:
shall we fail here if `needBroadcast = false`?

Reply:
Hm, is it necessary?

Reply:
maybe just adding an assert for sanity check would be fine?

Reply:
or combine w/ the `require` below

Reply:
Oh, seems we have different evaluation for DPP. `DataSourceScanExec` calls DPP filters' evaluation. But for v2 scan, we translate to source filters and let scan to do filtering internally, so it won't call catalyst expression evaluation.

So we cannot fail in `prepareResult` now.

Reply:
hmm, do you mean v2 scan DPP will still do broadcast?

Reply:
No, it evaluates locally.